src/bootstrap/config.rs

```diff
@@ -816,8 +816,16 @@ impl Config {
                 check_ci_llvm!(llvm.allow_old_toolchain);
                 check_ci_llvm!(llvm.polly);
 
-                // CI-built LLVM is shared
-                config.llvm_link_shared = true;
+                // CI-built LLVM can be either dynamic or static.
+                let ci_llvm_lib =
+                    config.out.join(&*config.build.triple).join("ci-llvm").join("lib");
+                config.llvm_link_shared =
+                    match std::fs::read_to_string(ci_llvm_lib.join("link-type.txt")) {
+                        Ok(link_type) => link_type == "dynamic",
+                        // If the file isn't found, this is probably before support for static linking was enabled.
+                        // Fall back to dynamic linking.
```

Comment:
We should never be in such a state though? So I'm a bit confused by this. I would prefer to error out.

I think this is probably better, but it's worth noting that in theory llvm-config can tell us the mode LLVM was built with (`llvm-config --shared-mode`).

Reply:
So my thought was that you could have a new version of bootstrap downloading an old version of ci-llvm. But I guess if you have a new version, it will always have gone through CI first, so there will be always be artifacts available. I'll change this to error out.

Reply:
Err actually I found the case where this fails - locally the first time I commit it, because there's no artifacts available yet:
```
thread 'main' panicked at 'std::fs::read_to_string(ci_llvm_lib.join("link-type.txt")) failed with No such file or directory (os error 2)', src/bootstrap/config.rs:822:33
```
```
> ls build/x86_64-unknown-linux-gnu/ci-llvm/lib/
libLLVM-11-rust-1.49.0-nightly.so  libLLVM-11-rust-1.51.0-nightly.so
libLLVM-11-rust-1.50.0-nightly.so
```
So up to you if I should keep this or not - if I don't keep it, I won't be able to test changes.

Reply:
Hm, I think given that we know this works, I would prefer you not keep it. You aren't actually testing whether it works right now, really, since you're using an old tarball - if you were using the new tarball (for example by switching to a try commit from this PR) then it should still work, right?

Reply:
I think so, yes. I can't test because none of the try builds were for linux.